tree/ntuple/v7/inc/ROOT/RPageStorageFile.hxx

```diff
@@ -0,0 +1,205 @@
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RPageStorageFile
+#define ROOT7_RPageStorageFile
+
+#include <ROOT/RPageStorage.hxx>
+#include <ROOT/RNTupleMetrics.hxx>
+#include <ROOT/RNTupleZip.hxx>
+#include <ROOT/RStringView.hxx>
+
+#include <array>
+#include <cstdio>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+namespace ROOT {
+namespace Experimental {
+
+// clang-format off
+/**
+\class ROOT::Experimental::RNTuple
+\ingroup NTuple
+\brief Entry point for an RNTuple in a ROOT file
+
+The class points to the header and footer keys, which in turn have the references to the pages.
+Only the RNTuple key will be listed in the list of keys. Like TBaskets, the pages are "invisible" keys.
+Byte offset references in the RNTuple header and footer reference directly the data part of page records,
+skipping the TFile key part.
+*/
+// clang-format on
+struct RNTuple {
+   /// Allows for evolving the struct in future versions
+   std::uint32_t fVersion = 0;
+   /// Allows for skipping the struct
+   std::uint32_t fSize = sizeof(RNTuple);
+   /// The file offset of the key containing the ntuple header
+   std::uint64_t fSeekHeader = 0;
+   /// The size of the compressed ntuple header including the TKey
+   std::uint32_t fNBytesHeader = 0;
+   /// The size of the uncompressed ntuple header
+   std::uint32_t fLenHeader = 0;
+   /// The file offset of the key containing the ntuple footer
+   std::uint64_t fSeekFooter = 0;
+   /// The size of the compressed ntuple footer including the TKey
+   std::uint32_t fNBytesFooter = 0;
+   /// The size of the uncompressed ntuple footer
+   std::uint32_t fLenFooter = 0;
+   /// Currently unused, reserved for later use
+   std::uint64_t fReserved = 0;
+};
+
+
+namespace Internal {
+/// Holds references to an open ROOT file during writing
+struct RTFileControlBlock;
+} // namespace Internal
+
+
+namespace Detail {
+
+class RPageAllocatorHeap;
+class RPagePool;
+class RRawFile;
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RPageSinkFile
+\ingroup NTuple
+\brief Storage provider that write ntuple pages into a file
+
+The written file can be either in ROOT format or in raw format.
+*/
+// clang-format on
+class RPageSinkFile : public RPageSink {
+private:
+   static constexpr std::size_t kDefaultElementsPerPage = 10000;
+
+   RNTupleMetrics fMetrics;
+   std::unique_ptr<RPageAllocatorHeap> fPageAllocator;
+
+   FILE *fFile = nullptr;
+   /// Byte offset of the next write (current file size)
+   std::uint64_t fFilePos = 0;
```

Comment:
Is ```fFilePos``` only used as the 'next' place to write in the file or also as the 'end of file location'?  If both concepts are used (and overlapped then allowing (later) 'freeing' and 'reuse' of space within the space will be a bit harder).  Also (much more relevant for reading then writing), we found out that recording the state of the file (when reading the "last" seek location) to be an hindrance in adding the ability of concurrent reads (eg. for a 'concurrent' TFile).

Reply:
This is now changed such that the file pos is only used internally by `RMiniFile` to remember the end of file when using a C file stream. The `RPageStorageFile` class, which uses `RMiniFile`, only gets access to the offset of a blob as the function result of writing. This is enough information to populate the ntuple meta-data. In particular, for determining the byte range of a cluster, the page source does not rely on the difference of file positions anymore but it uses the minimum and maximum offset of all the pages written for this cluster.